velox/functions/prestosql/types/IPPrefixType.cpp

```diff
@@ -0,0 +1,308 @@
+        // Take care of mask() throw condition
+        if (net.second > kIPV4Bits) {
+          context.setStatus(row, threadSkipErrorDetails() ? Status::UserError() : Status::UserError(
+                      "CIDR value '{}' is > network bit count '{}'",
+                      net.second,
+                      kIPV4Bits));
+          return;
+        }
+        addrBytes = folly::IPAddress::createIPv4(net.first)
+                        .mask(net.second)
+                        .createIPv6()
+                        .toByteArray();
+      } else {
+         // Take care of mask() throw condition
+        if (net.second > kIPV6Bits) {
+          context.setStatus(row, threadSkipErrorDetails() ? Status::UserError() : Status::UserError(
+                    "CIDR value '{}' is > network bit count '{}'",
+                    net.second,
+                    kIPV6Bits));
+          return;
+        }
+        addrBytes = folly::IPAddress::createIPv6(net.first)
+                        .mask(net.second)
+                        .toByteArray();
+      }
+
+      exec::StringWriter<false> result(flatResult, row);
+      result.resize(kIPPrefixBytes);
+      memcpy(result.data(), &addrBytes, kIPAddressBytes);
+      result.data()[kIPAddressBytes] = net.second;
+      result.finalize();
+
+    });
+  }
+
+  static void castToIPAddress(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      BaseVector& result) {
+    auto* flatResult = result.as<FlatVector<int128_t>>();
+    const auto* ipaddresses = input.as<SimpleVector<StringView>>();
+
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      const auto intAddr = ipaddresses->valueAt(row);
+      int128_t addrResult = 0;
+      folly::ByteArray16 addrBytes;
+
+      memcpy(&addrBytes, intAddr.data(), kIPAddressBytes);
+      std::reverse(addrBytes.begin(), addrBytes.end());
+
+      memcpy(&addrResult, &addrBytes, kIPAddressBytes);
+      flatResult->set(row, addrResult);
+    });
+  }
+
+  static void castFromIPAddress(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      BaseVector& result) {
+    auto* flatResult = result.as<FlatVector<StringView>>();
+    const auto* ipAddresses = input.as<SimpleVector<int128_t>>();
+
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      auto ipAddress = ipAddresses->valueAt(row);
+      folly::ByteArray16 addrBytes;
+
+      exec::StringWriter<false> result(flatResult, row);
+      result.resize(kIPPrefixBytes);
+
+      memcpy(&addrBytes, &ipAddress, kIPAddressBytes);
+      std::reverse(addrBytes.begin(), addrBytes.end());
+      memcpy(result.data(), &addrBytes, kIPAddressBytes);
+
+      folly::IPAddressV6 v6Addr(addrBytes);
+      if (v6Addr.isIPv4Mapped()) {
+        result.data()[kIPAddressBytes] = kIPV4Bits;
+      } else {
+        result.data()[kIPAddressBytes] = (unsigned char)kIPV6Bits;
```

Comment:
Is this cast needed? Both `kIPV4Bits` and `kIPV6Bits` so not sure what the difference is here.

Reply:
Update: Re-added unsigned char. Mac-os build check failed due to conversion from 128 -> -128.